### DISCUSSION TITLE
How to implement Kafka Serializers and Deserializers?

### DISCUSSION THREAD

**Discussion by giscus** [quarkusio/quarkus#39565] [Quarkus Blog/Website]
# How to implement Kafka Serializers and Deserializers?

Quarkus: Supersonic Subatomic Java

https://quarkus.io/blog/kafka-serde/

**Comment by adamkotwasinski** [comment]
> Custom serializers and deserializers cannot be CDI beans. Kafka instantiates them directly using reflection.

What effectively means no access to the rest of the framework in these (de)serializers. And no control over how many instances of these deserializers are created due to [Kafka doing it on its own](https://github.com/apache/kafka/blob/3.7.0/clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java#L395).

FWIW allows for injection of custom deserializer as a (shared) object through [one of its constructors](https://github.com/apache/kafka/blob/3.7.0/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L573). So that could be a workaround.

**Comment by adamkotwasinski** [comment]
Also, we are dealing with Kafka-level conversions, so in our receiver code we either work with `ConsumerRecord<KeyType, ValueType>` , `Message`, or `ValueType`.

How would one provide a simple (framework-aware, e.g. for metrics) converter that can take both key and value and pass them to the channel? 

Effectively it means putting some kind of conversion hook between the `consumer.poll()` result (so the Kafka-level deserialization is already done) and userspace (e.g. invocation of `@Incoming` annotated methods).
